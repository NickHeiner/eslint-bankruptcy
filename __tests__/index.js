const {dirSync: createTmpDir} = require('tmp');
const mkdirp = require('mkdirp');
const fs = require('fs');
const path = require('path');
// @ts-ignore TS erroneously complains about this require.
const packageJson = require('../package');
const {spawnSync} = require('child_process');
const globby = require('globby');
const log = require('nth-log');

/**
 * @param {string} testName 
 * @param {string[]} flagsOtherThanFilePath 
 * @param {string[]} [filesArgs]
 */
function prepareTest(testName, flagsOtherThanFilePath, filesArgs) {
  const tmpDir = createTmpDir({prefix: `${packageJson.name}-${encodeURIComponent(testName)}-`}).name;
  copy(path.resolve(__dirname, '..', 'fixtures'), tmpDir);
  const binPath = path.resolve(__dirname, '..', packageJson.bin['declare-eslint-bankruptcy']);
  const filesToPass = filesArgs ? filesArgs.map(filePath => path.join(tmpDir, filePath)) : [tmpDir];
  const flags = [...filesToPass, ...flagsOtherThanFilePath];
  log.trace({binPath, flags}, 'Spawning');
  const {status, stdout: stdoutBuffer, stderr: stderrBuffer} = spawnSync(binPath, flags, {
    env: {
      ...process.env,
      loglevel: 'trace'
    }
  });
  const stdout = stdoutBuffer.toString();
  const stderr = stderrBuffer.toString();
  if (process.env.loglevel === 'trace') {
    console.log('stdout', stdout, 'stderr', stderr);
  }
  if (status) {
    const err = new Error('Spawning declare-eslint-bankruptcy failed');
    Object.assign(err, {stdout, stderr});
    throw err;
  }
    
  return {
    files: filesArgs ? filesToPass : globby.sync(`${tmpDir}/**/*`), 
    rootDir: tmpDir
  };
}

/**
 * 
 * @param {string} sourceDir 
 * @param {string} destDir 
 */
function copy(sourceDir, destDir) {
  const files = globby.sync(['**/*', '**/.*'], {cwd: sourceDir});
  files.forEach(filePath => {
    const sourcePath = path.join(sourceDir, filePath);
    const destPath = path.join(destDir, filePath);
    log.trace({sourcePath, destPath}, 'Copying file');
    mkdirp.sync(path.dirname(destPath));
    fs.writeFileSync(destPath, fs.readFileSync(sourcePath, 'utf8'));
  });
}

/**
 * 
 * @param {object} options 
 * @param {string[]} options.files
 * @param {string} options.rootDir
 */
function assertFilesMatchSnapshots({files, rootDir}) {
  files.forEach(filePath => {
    it(path.relative(rootDir, filePath), () => {
      expect(fs.readFileSync(filePath, 'utf8')).toMatchSnapshot();
    });
  });
}

describe('eslint-bankruptcy', () => {
  describe('only no-console', () => {
    const files = prepareTest('only no-console', ['--rule', 'no-console']);
    assertFilesMatchSnapshots(files);
  });

  describe('multiple explicit file paths', () => {
    const files = prepareTest(
      'multiple explicit file paths', 
      ['--rule', 'no-console'], 
      ['root.js', 'already-disabled-line.js']
    );
    assertFilesMatchSnapshots(files);
  });

  describe('warnings are ignored', () => {
    const files = prepareTest('warnings are ignored', ['--rule', 'eqeqeq'], ['only-warning.js']);
    assertFilesMatchSnapshots(files);
  });
  
  describe('set explanation message', () => {
    const files = prepareTest('only no-console', ['--rule', 'no-console', '--explanation', 'inserted explanation']);
    assertFilesMatchSnapshots(files);
  });

  describe('no-console and camelcase', () => {
    const files = prepareTest('only no-console', ['--rule', 'no-console', '--rule', 'camelcase']);
    assertFilesMatchSnapshots(files);
  });

  describe('dry run', () => {
    const files = prepareTest('dry run', ['--dry-run', '--rule', '--no-console']);
    assertFilesMatchSnapshots(files);
  });
});