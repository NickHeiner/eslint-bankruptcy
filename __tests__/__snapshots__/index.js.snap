// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eslint-bankruptcy dry run already-disabled-file.js 1`] = `
"/* eslint-disable no-console */

function f() {}

console.log();
"
`;

exports[`eslint-bankruptcy dry run already-disabled-line.js 1`] = `
"console.log();

// Already disabled
// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy dry run multiple-rules-one-line.js 1`] = `
"let foo_bar;

console.log(); foo_bar;"
`;

exports[`eslint-bankruptcy dry run nest/nested.js 1`] = `
"console.log();
"
`;

exports[`eslint-bankruptcy dry run only-warning.js 1`] = `
"let x, y, f;
if (x == y) {
  f();
}"
`;

exports[`eslint-bankruptcy dry run root.js 1`] = `
"console.log();

// Already has comment
console.log();

function indented() {
  console.log();
}
"
`;

exports[`eslint-bankruptcy multiple explicit file paths already-disabled-line.js 1`] = `
"// eslint-disable-next-line no-console
console.log();

// Already disabled
// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy multiple explicit file paths root.js 1`] = `
"// eslint-disable-next-line no-console
console.log();

// Already has comment
// eslint-disable-next-line no-console
console.log();

function indented() {
  // eslint-disable-next-line no-console
  console.log();
}
"
`;

exports[`eslint-bankruptcy no-console and camelcase already-disabled-file.js 1`] = `
"/* eslint-disable no-console */

function f() {}

console.log();
"
`;

exports[`eslint-bankruptcy no-console and camelcase already-disabled-line.js 1`] = `
"// eslint-disable-next-line no-console
console.log();

// Already disabled
// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy no-console and camelcase multiple-rules-one-line.js 1`] = `
"// eslint-disable-next-line camelcase
let foo_bar;

// eslint-disable-next-line no-console camelcase
console.log(); foo_bar;"
`;

exports[`eslint-bankruptcy no-console and camelcase nest/nested.js 1`] = `
"// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy no-console and camelcase only-warning.js 1`] = `
"let x, y, f;
if (x == y) {
  f();
}"
`;

exports[`eslint-bankruptcy no-console and camelcase root.js 1`] = `
"// eslint-disable-next-line no-console
console.log();

// Already has comment
// eslint-disable-next-line no-console
console.log();

function indented() {
  // eslint-disable-next-line no-console
  console.log();
}
"
`;

exports[`eslint-bankruptcy only no-console already-disabled-file.js 1`] = `
"/* eslint-disable no-console */

function f() {}

console.log();
"
`;

exports[`eslint-bankruptcy only no-console already-disabled-line.js 1`] = `
"// eslint-disable-next-line no-console
console.log();

// Already disabled
// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy only no-console multiple-rules-one-line.js 1`] = `
"let foo_bar;

// eslint-disable-next-line no-console
console.log(); foo_bar;"
`;

exports[`eslint-bankruptcy only no-console nest/nested.js 1`] = `
"// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy only no-console only-warning.js 1`] = `
"let x, y, f;
if (x == y) {
  f();
}"
`;

exports[`eslint-bankruptcy only no-console root.js 1`] = `
"// eslint-disable-next-line no-console
console.log();

// Already has comment
// eslint-disable-next-line no-console
console.log();

function indented() {
  // eslint-disable-next-line no-console
  console.log();
}
"
`;

exports[`eslint-bankruptcy set explanation message already-disabled-file.js 1`] = `
"/* eslint-disable no-console */

function f() {}

console.log();
"
`;

exports[`eslint-bankruptcy set explanation message already-disabled-line.js 1`] = `
"// inserted explanation
// eslint-disable-next-line no-console
console.log();

// Already disabled
// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy set explanation message multiple-rules-one-line.js 1`] = `
"let foo_bar;

// inserted explanation
// eslint-disable-next-line no-console
console.log(); foo_bar;"
`;

exports[`eslint-bankruptcy set explanation message nest/nested.js 1`] = `
"// inserted explanation
// eslint-disable-next-line no-console
console.log();
"
`;

exports[`eslint-bankruptcy set explanation message only-warning.js 1`] = `
"let x, y, f;
if (x == y) {
  f();
}"
`;

exports[`eslint-bankruptcy set explanation message root.js 1`] = `
"// inserted explanation
// eslint-disable-next-line no-console
console.log();

// Already has comment
// inserted explanation
// eslint-disable-next-line no-console
console.log();

function indented() {
  // inserted explanation
  // eslint-disable-next-line no-console
  console.log();
}
"
`;

exports[`eslint-bankruptcy warnings are ignored only-warning.js 1`] = `
"let x, y, f;
if (x == y) {
  f();
}"
`;
